### PR TITLE
[NFC] Expose the TaskQueue as a Compilation Parameter

### PR DESCRIPTION
**Pull request by CodaFi**
In preparation for a future where a driver can specify its own TaskQueue, lift the responsibility for its creation to the driver and provide a convenient builder function to handle the argument parsing.

The second commit softens a condition that used to use `SkipTaskExecution` to avoid writing compilation records when the DummyTaskQueue is around.  In my tests it didn't make a difference. Nevertheless, the second commit is revertible if I'm wrong about that.

### CODE REVIEW

File: include/swift/Driver/Compilation.h [hunk 1]
```diff
@@ -330,9 +321,12 @@ class Compilation {
   const char *getAllSourcesPath() const;
 
   /// Asks the Compilation to perform the Jobs which it knows about.
+  ///
+  /// \param TaskQueue The TaskQueue used to schedule jobs for execution.
+  ///
   /// \returns result code for the Compilation's Jobs; 0 indicates success and
   /// -2 indicates that one of the Compilation's Jobs crashed during execution
-  int performJobs();
+  int performJobs(std::unique_ptr<sys::TaskQueue> &&TaskQueue);
```
Comment (davidungar) [inline include/swift/Driver/Compilation.h]: Two comments here:
1. Although it's fairly common practice in C++, I find it more readable to use a different identifier for the instance than for the type, perhaps `TheTaskQueue` or `TQ`?
2. Might a caller ever want to inspect the task queue after `performJobs` returns?

File: lib/Driver/Compilation.cpp [hunk 2]
```diff
@@ -889,7 +880,7 @@ namespace driver {
         return;
       }
 
-      size_t NumPartitions = Comp.NumberOfParallelCommands;
+      size_t NumPartitions = TQ->getNumberOfParallelTasks();
```
Comment (davidungar) [inline lib/Driver/Compilation.cpp]: Nice!

**Review by davidungar — COMMENTED** [review]
Just a few minor things...

File: include/swift/Driver/Compilation.h (hunk 1 above)
Comment (CodaFi) [reply include/swift/Driver/Compilation.h]: 1) Agreed

2) For my purposes and the purposes of the driver, both only need the `TaskQueue` to live as long as the Compilation session.  @graydon or @jrose-apple can probably think of concrete reasons why we may want to extend it in the future.

File: tools/driver/driver.cpp [hunk 3]
```diff
@@ -219,11 +220,14 @@ int main(int argc_, const char **argv_) {
...
   std::unique_ptr<Compilation> C =
       TheDriver.buildCompilation(*TC, std::move(ArgList));
+
   if (Diags.hadAnyError())
     return 1;
 
   if (C) {
-    return C->performJobs();
+    std::unique_ptr<sys::TaskQueue> TQ =
+      TheDriver.buildTaskQueue(*C, C->getArgs());
```
[the remark this answers is not in the document]
Comment (CodaFi) [reply tools/driver/driver.cpp]: Good catch!

**Review by graydon — APPROVED** [review]
LGTM!

**Comment by CodaFi** [comment]
Going to merge this so @davidungar can rebase his patch onto these changes.

⛵️

**Comment by CodaFi** [comment]
@swift-ci please test and merge

File: include/swift/Driver/Compilation.h (hunk 1 above)
Comment (jrose-apple) [reply include/swift/Driver/Compilation.h]: I think to David's point there's no reason to have the Compilation *destroy* the TaskQueue, although it is going to irrevocably do a bunch of stuff to it. An lvalue reference (or a normal pointer) probably makes more sense.

**Pull request merged**